Blog: update of the OpenLLMLeaderboard (#1666)
huggingface/blog: leaderboard-drop-dive.md

---
title: "DROP deep dive"
thumbnail: /blog/assets/evaluating-mmlu-leaderboard/thumbnail.png
authors:
- user: clefourrier
- user: cabreraalex
- user: stellaathena
- user: SaylorTwift
- user: thomwolf
---

# DROP deep dive
Recently, [three new benchmarks](https://twitter.com/clefourrier/status/1722555555338956840) were added to the [Open LLM Leaderboard](https://huggingface.co/spaces/HuggingFaceH4/open_llm_leaderboard): Winogrande, GSM8k and DROP. A cursory look at the scores for DROP revealed something strange was going on, with the overwhelming majority of models scoring less than 10 out of 100 on their f1-score! We did a deep dive to understand what was going on, come with us to see what we found out!

## Initial observations
DROP (Discrete Reasoning Over Paragraphs) is an evaluation where models must extract relevant information from English-text paragraphs before executing discrete reasoning steps on them (for example, sorting or counting items to arrive at the correct answer, see the table below for examples). The metrics used are custom f1 and exact match scores.

<div align="center">
<figure class="image table text-center m-0 w-full">

>   <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/open-llm-leaderboard/drop/drop_example.png" width="500" />

Author: I can't manage to push there, I'll try to manually upload them

  <figcaption>Examples of reasoning and paragraph from the original article.</figcaption>
</figure>
</div>

We added it to the Open LLM Leaderboard three weeks ago, and observed that the f1-scores of pretrained models followed an unexpected trend: when we plotted DROP scores against the leaderboard original average (of ARC, HellaSwag, TruthfulQA and MMLU), which is a reasonable proxy for overall model performance, we expected DROP scores to be correlated with it (with better models having better performance). However, this was only the case for a small number of models, and all the others had a very low DROP f1-score, below 10.

<div align="center">
<figure class="image table text-center m-0 w-full">
  <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/open-llm-leaderboard/drop/drop_bimodal.png" width="500" />
  <figcaption>Two trends can be observed in the DROP scores: some follow the average (in diagonal), others are stuck around 5 (vertical line on the right of the graph).</figcaption>
</figure>
</div>

## Normalization interrogations

> During preliminary observations, the leaderboard team observed that the normalization was not working as expected: in some cases, it ignored the correct numerical answers when they were directly followed by a whitespace which is not a space (such as a line return). 

Reviewer: ```suggestion
During preliminary observations, the leaderboard team observed that the normalization step was not working as expected: in some cases, it ignored the correct numerical answers when they were directly followed by a whitespace character other than a space (a line return, for example).
```

Let's look at an example, with the generation being `10\n\nPassage: The 2011 census recorded a population of 1,001,360`, and the gold answer being `10`.

Normalization happens in several steps, both for generation and gold:

> 1) **Tokenization on separators** `|`, `-`, or `.` 

Reviewer: ```suggestion
1) **Split on separators** `|`, `-`, or `.` 
```

nit: IIUC these are not tokenizer tokens but substrings, is that right?

Author: Yep

4) **Other steps**
  A lot of other normalization steps ensue (removing articles, removing other whitespaces, etc.) and our original example becomes `10 passage 2011.0 census recorded population of 1001360.0`.

However, the overall score is not computed on the string, but on the bag of words (BOW) extracted from the string, here `{'recorded', 'population', 'passage', 'census', '2011.0', '1001360.0', '10'}`, which is compared with the BOW of the gold, also normalized in the above manner, `{10.0}`. As you can see, they don’t intersect, even though the model predicted the correct output!

In summary, if a number is followed with any kind of whitespace other than a simple space, it will not pass through the number normalization, hence never match the gold if it is also a number! This first issue was likely to mess up the scores quite a bit, but was also clearly not the only factor at hand for DROP scores so low, so we decided to investigate a bit more.

## Diving into the results

> Our friends at Zeno [took a deeper dive](https://hub.zenoml.com/report/1255/DROP%20Benchmark%20Exploration) in the detailed results, looking at 5 models which were representative of the problems encountered in DROP scores: falcon-180B and mistral-7B were underperforming compared to what we were expecting, Yi-34B and tigerbot-70B had a very good performance on DROP correlated with their average scores, and facebook/xglm-7.5B fell in the middle. 

Reviewer: ```suggestion
Our friends at Zeno [took a deeper dive](https://hub.zenoml.com/report/1255/DROP%20Benchmark%20Exploration) over the detailed results, looking at 5 models which were representative of the problems encountered in DROP scores: falcon-180B and mistral-7B were underperforming compared to what we were expecting, Yi-34B and tigerbot-70B had a very good performance on DROP correlated with their average scores, and facebook/xglm-7.5B fell in the middle. 
```

They found out two additional interesting points:
1) Not a single model got a correct result on floating point answers
2) High quality models which generate long answers actually have a lower f1-score

After this investigation, we believed that both these points are actually caused by the same root factor: using `.` as stopword token (to end the generations):
1) Floating point answers are systematically interrupted before their generation is complete
2) Higher quality models, which try to match the few-shot prompt format, will generate `Answer\n\nPlausible prompt for the next question.`, and only stop during the plausible prompt continuation after the actual answer on the first `.`, therefore generating too many words and getting a bad f1 score.

We thought both these problems would likely be fixed by using `\n` instead of `.` as an end of generation stop word.

## Changing the end of generation token
So we gave it a try! We investigated using `\n` as the end of generation token on the available results. We split the generated answer on the first `\n` it contained, if one was present, and recomputed the scores. 

> *Note that this is only an approximation of the correct result, as it won't fix answers that were cut too early on `.` (for example floating point answers) - but it also won’t advantage any model, as all of them were affected by this problem. 

Reviewer: Ahh, yes. Thank you!

  <figcaption>We can see in orange that the scores computed on the new strings correlate much better with the average performance.</figcaption>
</figure>
</div>

## So what next?
As re-running the full evaluation of all models would be quite costly (the full update took 8 years of GPU time, and a lot of it was taken by DROP), we estimated how much it would cost to only re-run failing examples.

In 10% of cases, the gold is a floating number (for example `12.25`) and the model prediction starts with the correct beginning (for our example, `12`) but was cut off on a `.` - these generations are very likely to have actually been correct if they could have finished. We would definitely need to re-run them! 
This estimation does not count generated sentences that finish with a number which was possibly interrupted (40% of the other generations), nor any prediction messed up by its normalization.

To get correct results, we would possibly need to re-run more than 50% of the examples, which represents a huge amount of GPU time: we are talking several months of continuous computations! So we need to be certain that the implementation is correct this time.

> After talking it over with the fantastic EleutherAI team (both on [GitHub](https://github.com/EleutherAI/lm-evaluation-harness/issues/978) and internally), who guided us through the code and helped our investigations, it became very clear that the LM Eval Harness implementation follows the official DROP code strictly: a new version of this benchmark’s evaluation needs to be developed! 

Reviewer: I'd maybe mention the offical code (and Eleuther's implementation) earlier, to explain this is the reason why we were using the normalization rules you described :)

Reviewer 2: I agree. In one draft it was at the end of "Initial observations" but I think it got bumped here because it's also related to the convo about next steps.

**We have therefore taken the decision to remove DROP from the Open LLM Leaderboard until a new version arises.**

> We hope that interested members of the community will join forces with academics working on evaluation to fix both its scoring and its normalization, to make it once again usable, as the dataset itself is really quite interesting and cool. We encourage you to provide feedback on how we should evaluate DROP [on this issue](https://github.com/EleutherAI/lm-evaluation-harness/issues/1050).

Reviewer: ```suggestion
We hope that interested members of the community will join forces with academics working on DROP evaluation to fix both its scoring and its normalization. We'd love it becomes usable again, as the dataset itself is really quite interesting and cool. We encourage you to provide feedback on how we should evaluate DROP [on this issue](https://github.com/EleutherAI/lm-evaluation-harness/issues/1050).
```

Reviewer: Maybe add another point in the conclusion about how delicate and difficult evaluation is :)

This estimation does not count generated sentences that finish with a number which was possibly interrupted (40% of the other generations), nor any prediction messed up by its normalization.

To get correct results, we would possibly need to re-run more than 50% of the examples, which represents a huge amount of GPU time: we are talking several months of continuous computations! So we need to be certain that the implementation is correct this time.

After talking it over with the fantastic EleutherAI team (both on [GitHub](https://github.com/EleutherAI/lm-evaluation-harness/issues/978) and internally), who guided us through the code and helped our investigations, it became very clear that the LM Eval Harness implementation follows the official DROP code strictly: a new version of this benchmark’s evaluation needs to be developed! 
**We have therefore taken the decision to remove DROP from the Open LLM Leaderboard until a new version arises.**

> We hope that interested members of the community will join forces with academics working on DROP evaluation to fix both its scoring and its normalization. We'd love it becomes usable again, as the dataset itself is really quite interesting and cool. We encourage you to provide feedback on how we should evaluate DROP [on this issue](https://github.com/EleutherAI/lm-evaluation-harness/issues/1050).

Author: Add mention of how delicate eval is
